File: src/types/accounts.ts
import type { Syncable } from './common'

export type AccountType = 'cash' | 'bank' | 'credit' | 'ewallet' | 'investment' | 'debt' | 'receivable' | 'other'

export interface Account extends Syncable {
  id: string
  name: string
  type: AccountType
  currency_code: string
  initial_balance_cents: number
  created_at: string
  is_hidden: boolean
}

export interface AccountInput {
  name: string
  type: AccountType
  currency_code: string
  initial_balance_cents?: number
}

/** 账户编辑入参：type 不可改（参与余额符号归属）；币种仅无交易账户可改（后端拒绝）。 */
export interface AccountUpdateInput {
  name?: string
  currency_code?: string
}

/** 余额调整入参：校准到目标值，后端生成一笔与黑洞账户的转账（ADR-0026）。 */
export interface AccountBalanceAdjustInput {
  target_balance_cents: number
  date: string
  note?: string
}

export interface AccountBalance {
  account: Account
  balance_cents: number
}

/** 余额缓存审计差异行（issue #491）：缓存缺失记 null（回填前）。 */
export interface BalanceCacheDrift {
  account_id: string
  account_name: string
  cached_cents: number | null
  actual_cents: number
}

/** 余额缓存审计报告（issue #491）：修复已完成后的差异快照。 */
export interface BalanceCacheAudit {
  accounts_checked: number
  drifts: BalanceCacheDrift[]
  repaired: boolean
}

/** 账户类型闭集（穷尽表驱动：新增 AccountType 变体未列出即编译报错）；
 * 显示标签在文案资源 accounts.type.*（i18n，ADR-0049），不再硬编码。 */
const ACCOUNT_TYPE_PRESENCE = {
  cash: true,
  bank: true,
  credit: true,
  ewallet: true,
  investment: true,
  debt: true,
  receivable: true,
  other: true,
} satisfies Record<AccountType, boolean>

export const ACCOUNT_TYPES = Object.keys(ACCOUNT_TYPE_PRESENCE) as AccountType[]

/** 出资账户准入闭集（issue #936 / ADR-0096 决策 4）：现金类账户 cash/bank/credit/ewallet/other。
 * 排除 investment（跨投资账户走 transfer，保住「子弹」语义单一）与 receivable/debt（借贷
 * 账户不承载投资结算）；与后端 transaction/funding.rs 的 FUNDING_ALLOWED_TYPES 同一闭集，
 * 后端行为层准入是唯一权威，本表驱动谓词仅供表单候选预过滤。币种一致过滤在表单层
 * （随交易币种），不在此闭集内。 */
const FUNDING_ACCOUNT_TYPE_PRESENCE = {
  cash: true,
  bank: true,
  credit: true,
  ewallet: true,
  other: true,
} satisfies Partial<Record<AccountType, boolean>>

/** 出资账户候选谓词：账户类型在准入闭集内（币种一致过滤由调用方按交易币种承担） */
export function isFundingCandidateAccount(type: AccountType): boolean {
  return (FUNDING_ACCOUNT_TYPE_PRESENCE as Record<AccountType, boolean | undefined>)[type] === true
}

File: src-tauri/src/models/accounts.rs
//! 账户领域模型：账户类型枚举、账户实体与入参、账户余额。

use std::fmt;
use std::str::FromStr;

use rusqlite::types::{FromSql, FromSqlError, ToSql, ToSqlOutput, ValueRef};
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

use crate::db::query::FromRow;
use crate::error::AppError;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, ToSchema)]
#[serde(rename_all = "snake_case")]
pub enum AccountType {
    Cash,
    Bank,
    Credit,
    Ewallet,
    Investment,
    Debt,
    Receivable,
    Other,
}

impl fmt::Display for AccountType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            AccountType::Cash => write!(f, "cash"),
            AccountType::Bank => write!(f, "bank"),
            AccountType::Credit => write!(f, "credit"),
            AccountType::Ewallet => write!(f, "ewallet"),
            AccountType::Investment => write!(f, "investment"),
            AccountType::Debt => write!(f, "debt"),
            AccountType::Receivable => write!(f, "receivable"),
            AccountType::Other => write!(f, "other"),
        }
    }
}

impl FromStr for AccountType {
    type Err = AppError;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "cash" => Ok(AccountType::Cash),
            "bank" => Ok(AccountType::Bank),
            "credit" => Ok(AccountType::Credit),
            "ewallet" => Ok(AccountType::Ewallet),
            "investment" => Ok(AccountType::Investment),
            "debt" => Ok(AccountType::Debt),
            "receivable" => Ok(AccountType::Receivable),
            "other" => Ok(AccountType::Other),
            _ => Err(AppError::Invalid(format!("未知账户类型: {s}"))),
        }
    }
}

impl ToSql for AccountType {
    fn to_sql(&self) -> rusqlite::Result<ToSqlOutput<'_>> {
        Ok(ToSqlOutput::from(self.to_string()))
    }
}

impl FromSql for AccountType {
    fn column_result(value: ValueRef<'_>) -> std::result::Result<Self, FromSqlError> {
        value
            .as_str()?
            .parse()
            .map_err(|e: AppError| FromSqlError::Other(Box::new(e)))
    }
}

#[derive(Debug, Serialize, Deserialize, Clone, ToSchema)]
pub struct Account {
    pub id: String,
    pub name: String,
    #[serde(rename = "type")]
    pub kind: AccountType,
    pub currency_code: String,
    pub initial_balance_cents: i64,
    pub created_at: String,
    pub updated_at: String,
    pub version: i64,
    pub device_id: String,
    pub is_deleted: bool,
    /// 黑洞账户标志：对用户侧列表/余额/下拉选择器隐藏，但交易仍参与交易列表与报表。
    pub is_hidden: bool,
}

#[derive(Debug, Deserialize, ToSchema)]
pub struct AccountInput {
    pub name: String,
    #[serde(rename = "type")]
    pub kind: AccountType,
    pub currency_code: String,
    pub initial_balance_cents: Option<i64>,
}

/// 账户编辑入参（IPC `update_account` / HTTP `PUT /api/v1/accounts/{id}`）。
/// `type` 不可改：账户类型参与 kind→符号矩阵（余额方向），改动会重写历史交易
/// 的余额归属（ADR-0026 同期决策，Q3）；`initial_balance_cents` 不在此改，
/// 归余额调整（见参考数据与设置域 BalanceAdjustment）。
#[derive(Debug, Deserialize, ToSchema)]
pub struct AccountUpdateInput {
    pub name: Option<String>,
    /// 仅无交易账户可改（有交易时改币种会使历史折算口径错乱，后端拒绝）。
    pub currency_code: Option<String>,
}

/// 余额调整入参（IPC `adjust_account_balance`）：把余额校准到目标值，
/// 机制为生成一笔与黑洞账户的转账（ADR-0026）。
#[derive(Debug, Deserialize)]
pub struct AccountBalanceAdjustInput {
    pub target_balance_cents: i64,
    /// 调整交易日期（YYYY-MM-DD，对账常补记过去日期）。
    pub date: String,
    /// 调整交易备注；缺省后端补「余额调整」。
    pub note: Option<String>,
}

#[derive(Debug, Serialize, ToSchema)]
pub struct AccountBalance {
    pub account: Account,
    pub balance_cents: i64,
}

impl FromRow for Account {
    fn from_row(row: &rusqlite::Row) -> rusqlite::Result<Self> {
        Ok(Account {
            id: row.get(0)?,
            name: row.get(1)?,
            kind: row.get(2)?,
            currency_code: row.get(3)?,
            initial_balance_cents: row.get(4)?,
            created_at: row.get(5)?,
            updated_at: row.get(6)?,
            version: row.get(7)?,
            device_id: row.get(8)?,
            is_deleted: row.get::<_, i64>(9)? != 0,
            is_hidden: row.get::<_, i64>(10)? != 0,
        })
    }
}
